### ISSUE TITLE
[REVIEW]: Jabberwocky: an ontology-aware toolkit for manipulating text

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#2168]
**Submitting author:** @sap218 (<a href="http://orcid.org/0000-0002-6169-0135">Samantha Pendleton</a>)
**Repository:** <a href="https://github.com/sap218/jabberwocky" target ="_blank">https://github.com/sap218/jabberwocky</a>
**Version:** v1.0.0.0
**Editor:** @majensen
**Reviewer:** @wdduncan, @balhoff
**Archive:** <a href="https://doi.org/10.5281/zenodo.3922261" target="_blank">10.5281/zenodo.3922261</a>

## Status

[![status](https://joss.theoj.org/papers/10502055165be490feaa389d51fe99d6/status.svg)](https://joss.theoj.org/papers/10502055165be490feaa389d51fe99d6)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/10502055165be490feaa389d51fe99d6"><img src="https://joss.theoj.org/papers/10502055165be490feaa389d51fe99d6/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/10502055165be490feaa389d51fe99d6/status.svg)](https://joss.theoj.org/papers/10502055165be490feaa389d51fe99d6)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/sap218/jabberwocky" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@wdduncan & @balhoff, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @majensen know.

✨ **Please try and complete your review in the next two weeks** ✨ 

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/sap218/jabberwocky">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@sap218) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/sap218/jabberwocky">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@sap218) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @wdduncan, @balhoff it looks like you're currently assigned to review this paper :tada:.

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02168/joss.02168/10.21105.joss.02168.pdf)

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- None

MISSING DOIs

- https://doi.org/10.1093/database/bau033 may be missing for title: tagtog: interactive and text-mining-assisted annotation of gene mentions in PLOS full-text articles
- https://doi.org/10.3115/v1/p14-5010 may be missing for title: The Stanford CoreNLP natural language processing toolkit
- https://doi.org/10.1093/nar/gkr972 may be missing for title: Disease Ontology: a backbone for disease semantic integration
- https://doi.org/10.1016/j.ajhg.2008.09.017 may be missing for title: The Human Phenotype Ontology: a tool for annotating and analyzing human hereditary disease
- https://doi.org/10.1093/bib/bbv011 may be missing for title: The role of ontologies in biological and biomedical research: a functional perspective

INVALID DOIs

- None
```

**Comment by majensen** [comment]
Quick thing if you can @sap218 - can you add the URLs that whedon-the-bot found [above](https://github.com/openjournals/joss-reviews/issues/2168#issuecomment-601297454) to your refs?

**Comment by sap218** [comment]
Currently I have (as an example for tagtog):

```
@ARTICLE{Cejuela2014-lv,
  title    = "tagtog: interactive and text-mining-assisted annotation of gene
              mentions in {PLOS} full-text articles",
  author = ..........,
  url      = {https://doi.org/10.1093/database/bau033}
}
```

and it prints as "Retrieved from: https://doi.org/10.1093/database/bau033"

is "url" the correct bibtex key?

@majensen

**Comment by arfon** [comment]
@sap218 - please change them to keys like this:

```
doi  = "10.1093/database/bau033"
```

i.e. change the `url` to `doi` and strip off the `https://doi.org/` preamble.

**Comment by majensen** [comment]
Thanks @arfon

**Comment by sap218** [comment]
Thanks @arfon - @majensen I believe I have fixed the DOI issue now

**Comment by arfon** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.1093/database/bau033 is OK
- 10.3115/v1/p14-5010 is OK
- 10.1093/nar/gkr972 is OK
- 10.1016/j.ajhg.2008.09.017 is OK
- 10.1093/bib/bbv011 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by wdduncan** [comment]
I finished the checklist.

Comments:  

1. I could not install the repo as per the instructions. When running the install script, I received error:
```
ERROR: TEST FAILED: /Users/wdduncan/.local/lib/python3.7/site-packages/ does NOT support .pth files 
error: bad install directory or PYTHONPATH
```
This could be b/c I am using OS X, or perhaps b/c I installed using venv. 

2.. The examples reference a file named `/ontology/pocketmonsters.owl`. But this file is not in the repo.

The github documentation would benefit from having examples of expected output and descriptions of what the output means.

The paper refers to the jabberwocky-tests as a way to see the software at work. But that directory contains no instructions.

How long is the software supposed paper to be?

**Comment by majensen** [comment]
@sap218 please have a look at above issues of @wdduncan 
@wdduncan - paper can be long or short, as long as it covers the requirements in the "Software paper" section of the review 
thanks!

**Comment by danielskatz** [comment]
Well, actually, a goal of JOSS is that papers should be reasonably short. In https://joss.readthedocs.io/en/latest/submitting.html#what-should-my-paper-contain, we say, "the paper should be between 250-1000 words" though some papers are somewhat longer.

**Comment by majensen** [comment]
Thanks for the clarification @danielskatz!

**Comment by sap218** [comment]
@wdduncan @majensen will get on looking at the issues stated as soon as I can! Thanks.

**Comment by arfon** [comment]
:wave: @balhoff - just a friendly check-in to see how things are going with your review?

**Comment by balhoff** [comment]
Thanks for the reminder, @arfon! I can work on this this week.

**Comment by sap218** [comment]
@wdduncan @majensen @balhoff I have made some fixes/changes, particularly changes to fix @wdduncan comments - paper.md has been updated too for a little more clarification. Let me know if you need any help!

**Comment by sap218** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02168/joss.02168/10.21105.joss.02168.pdf)

**Comment by majensen** [comment]
@wdduncan @balhoff Can I get an update from you guys here? I'm hoping we're closing in on it. Thanks!

**Comment by balhoff** [comment]
Hi @majensen I ran into a blocker with https://github.com/sap218/jabberwocky/issues/10. I reported it in the tool repo, but probably should have clarified here that I was waiting on that before being able to proceed further.

**Comment by wdduncan** [comment]
Hi @majensen I was waiting to see if the issues referenced by @balhoff and myself had been fixed. I can go ahead and retest, though.

**Comment by sap218** [comment]
I'll take a look at those issues @balhoff @wdduncan again - (sorry for delay). Will update here when I can! :-)

**Comment by sap218** [comment]
Just an update - I fixed the issues on the repo if you wanted to retry :-) 
(turns out i did indeed have a missing dependency - really sorry about that)

**Comment by majensen** [comment]
@balhoff @wdduncan - any movement on checking Sam's work? thanks

**Comment by balhoff** [comment]
Yep it's working for me now; I'll continue the review.

**Comment by wdduncan** [comment]
@sap218 
I tried installing the software, but I am having issues. Steps so far:
1. Create venv using `python3 -m venv .env`, and active environment (works fine).
2. Run `pip install -r requirements.txt` (works fine).

But, when I run `python3 setup.py install --user`, I get message/error:
```
You are attempting to install a package to a directory that is not
on PYTHONPATH and which Python does not read ".pth" files from.  The
installation directory you specified (via --install-dir, --prefix, or
the distutils default setting) was:

    /Users/wdduncan/.local/lib/python3.7/site-packages/

and your PYTHONPATH environment variable currently contains:

    ''
```
I fixed this by removing `--user`. Perhaps it is related to my using a virtual environment? if so, it would be good to document this.

I ran the `jabberwocky-tests`.   

For the `catch` test, I am unsure what the output in `catch_output.txt` means. Did `catch` find the lines in the blog that contained terms from the ontology?  
Some documentation on what the output means would be helpful.

The `arise` test ran successfully. But, like the `catch` test, documentation would be helpful for describing the output.

The `arise` test:
```
arise -ontology ../ontology/pocketmonsters.owl -tfidf new_synonyms_tfidf.csv $ arise -o ../ontology/pocketmonsters.owl -f new_synonyms_tfidf.csv
```
 failed with error:
```
Usage: arise [OPTIONS]
Try 'arise --help' for help.

Error: no such option: -t
```
Hope this helps!

cc @majensen

**Comment by sap218** [comment]
@wdduncan Thank you so much! I will improve the documentation for users to understand, I see the problem with the `arise` failure (I made a minor typo in the documentation, I will fix that also). Thank you again!

**Comment by majensen** [comment]
Just a qn @wdduncan - is it weird to use --user when installing into a virtual environment? I wouldn't think to do it, since the environment is isolated from everything to begin with. (I only have toddler python chops, however.)

**Comment by wdduncan** [comment]
@majensen I'm not sure, I usually don't use the `--user` option with `pip`, unless I run into permission issues. Since, I was testing the software, I didn't want to affect the library I have. My guess is that in the venv, there isn't a "user" specified. So, `--user` doesn't make sense.  But, I may be wrong about this ... I'm not a virtual environment guru :).  

In any case, it is probably best document the behavior.

**Comment by majensen** [comment]
Ah! Well I believe the virtual environment is truly isolated from both your system install and your user packages. Everything is stored in the venv, including the python binary, pip and everything else. But - better safe than sorry. 
With python 3.7.6 (in an active venv) I get a more straightforward message:
`jensenma$ source venv/bin/activate
(venv) jensenma$ pip install --user requests
ERROR: Can not perform a '--user' install. User site-packages are not visible in this virtualenv.`

So I think it's a python thing, not a jabberwocky thing.

**Comment by wdduncan** [comment]
Yes, you right @majensen!  
I didn't mean to imply it was a jabberwocky thing. Only that that the installation directions might need to be updated with info about using venv in case someone tries to install in a virtual environment, gets an error and quits w/o further investigating the error.

**Comment by sap218** [comment]
@wdduncan I fixed that `arise` issue so you should be able to run that command now :-) I will be improving the documentation today

**Comment by sap218** [comment]
@wdduncan i have also improved the documentation to include explanations of the outputs, see the [SCENARIO.md](https://github.com/sap218/jabberwocky-tests/blob/master/SCENARIO.md)

**Comment by majensen** [comment]
@balhoff @wdduncan - how is it coming? @sap218 has put in the work so maybe we can get across the finish line.

**Comment by wdduncan** [comment]
Sorry ... did get a chance to review the changes this weekend. I will circle back around soon.

**Comment by balhoff** [comment]
Same here!

**Comment by wdduncan** [comment]
I did another run through the documentation on github. It is looking much better! I really like the figure at the bottom :)

However, I still have issues/comments:
* In the help section on https://github.com/sap218/jabberwocky there is a typo in: 
`-t, --textfile TEXT   **JSON ot TXT** file of text you want annotate. [required]` --> JSON **or** TXT file

* In the [scenario](https://github.com/sap218/jabberwocky-tests/blob/master/SCENARIO.md)
```
catch -o ../ontology/pocketmonsters.owl -k tests/catch/listofwords.txt -t public_forum.json -p post > catch_01_output.txt
```  
`public_forum.json` is not in the catch directory; I modified the command to point to bite/public_forum.js

* In the command:
```
bite -t public_forum.json -p post
```
What does the `-p post` do? I couldn't find documentation about it. Sorry, if I missed it.

* When creating the table for `arise`, is there an ontology or controlled vocabulary that is supposed to define the values in the "type" column?

I looked at the documentation for [rdfs:schema](https://www.w3.org/TR/rdf-schema/) 
But I did not find the annotation properties: rdfs:hasBroadSynonym, rdfs:hasNarrowSynonym, etc.

I also looked at the base IRI for [rdf-schema:hasBroadSynonym](https://www.w3.org/2000/01/rdf-schema#hasBroadSynonym) (i.e., https://www.w3.org/2000/01/rdf-schema) and still not find the aforementioned annotation properties.

Have you considered using [SKOS](https://www.w3.org/2004/02/skos/) which already has terms for broad match, close match, etc.

**Comment by sap218** [comment]
@wdduncan I looked over your comments:

* I fixed that typo "ot".. now fixed to "or"
* You mentioned that when looking at scenario.md that public_forum.json is not in the catch directory, but the scenario states at the top, "we are running these command in the jabberwocky-test/process directory" 
* I mention in the documentation that "-p, --parameter TEXT  parameter for the JSON file text." 
* the 'type' column is a person's own defined synonyms - I have just fixed this confusion
* I will look into the rdfs:synonyms annotations

I hope this helps!

**Comment by wdduncan** [comment]
@sap218 
> "we are running these command in the jabberwocky-test/process directory".   

Sorry. I missed this :)

> "-p, --parameter TEXT parameter for the JSON file text."   

Yes, I saw that. Sorry, I wasn't clear. What does the `post` mean? Is it different that `post_text`? How is it different than `-p user-text`? E.g.: `bite --textfile social_media_posts.json --parameter user-text`.  Is there a controlled set of terms you use with `-p` flag?

**Comment by sap218** [comment]
@wdduncan 

`-p` refers to the field inside the JSON file for the text data, "-p post_text" and "-p user-text" are just two different examples - sorry about this confusion, i will change the README so all examples only refer to one example.

**Comment by sap218** [comment]
@wdduncan I have just fixed the annotation issue regarding the synonyms, I fixed the documentation to explain this: 

> make sure annotations are defined with the oboInOWL: schema, e.g. hasExactSynonym should have the IRI http://www.geneontology.org/formats/oboInOWL#hasExactSynonym.

**Comment by wdduncan** [comment]
@sap218 That's helpful. I think letting the users define custom relations is fine too. You just have to be clear about what is going on. 

Here's a mapping project you might be interested: https://github.com/OBOFoundry/SSSOM

**Comment by sap218** [comment]
@wdduncan Great! Thank you! Glad I was able to address your concerns! :-)

**Comment by majensen** [comment]
> @wdduncan i have also improved the documentation to include explanations of the outputs, see the [SCENARIO.md](https://github.com/sap218/jabberwocky-tests/blob/master/SCENARIO.md)

(This is super cool.)

**Comment by wdduncan** [comment]
The documentation is looking good!

**Comment by majensen** [comment]
As I indicated to @wdduncan on another channel - just wanted to let all know that, as productive as it may be,  the review process _can_ be completed: When the reviewers finish their work, please check all the boxes in the review sheets and give us a heads-up with a comment here in the thread.

**Comment by majensen** [comment]
@balhoff - I see only one box unchecked in your review, Functionality. Are you close to being satisfied with that aspect?

@wdduncan - You have installion, functionality, and several documentation issue remaining. How are these coming?

This review began on March 19, and while there has been a lot of extra time provided due to covid, I think Sam has done a lot of good work improving the code and the MS. Time to draw a line under it. 
Thanks --@majensen

**Comment by wdduncan** [comment]
@majensen 
I checked boxes for installation and functionality. There aren't automated tests, but there is a tutorial. Are automated tests required?

@sap218 
A number of changes have been made to the github pages. Does the manuscript reflect these changes? I haven't had a chance to re-read it yet.

**Comment by majensen** [comment]
@wdduncan thanks - The tutorial can be used to verify functionality manually, so I think the letter of the law is satisfied. That said, @sap218 would you be willing to consider adding a few tests using [unittest](https://docs.python.org/3/library/unittest.html) or [pytest](https://docs.pytest.org/en/latest/index.html)? It is an important skill to have, if you haven't done it before, and it makes future support so much easier.

**Comment by sap218** [comment]
@wdduncan The manuscript still reflects the software 

@wdduncan @majensen 

thank you all for the comments throughout and continued patience. I do understand the value of unit testing and plan to explore this aspect of software development when improving the package in the near future, but I am currently constrained by time because of my continued PhD work, which makes it difficult for me to develop this testing skill and update the package - but I hope everyone would agree that the testing repository (and scenario.md) provided does allow users to check the package works as intended (as we have done so for this review) and has sufficient steps describing the functionality of the software.

If the reviewers believe that adding unit tests on top of the system test provided by the test repository should be a pre-requisite of approving the manuscript then I will do my best to invest the time.

**Comment by majensen** [comment]
@sap218 I understand and I will not hold up a recommendation on this basis.

**Comment by sap218** [comment]
@majensen  Thank you! I appreciate it :-)

**Comment by balhoff** [comment]
@majensen @sap218 I agree that the test/example repository is a fine functionality test; so I checked the tests box. I checked the last box, so I approve. Thank you both for your patience!

**Comment by majensen** [comment]
@whedon generate proof

**Comment by whedon** [comment]
I'm sorry human, I don't understand that. You can see what commands I support by typing:

```
@whedon commands
```

**Comment by majensen** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02168/joss.02168/10.21105.joss.02168.pdf)

**Comment by majensen** [comment]
Hey @sap218 - I added tests to jabberwocky; see [PR#13](https://github.com/sap218/jabberwocky/pull/13). These basically run the steps in the SCENARIO automatically, using the files in jabberwocky-tests. It's pretty basic, but now you have a working skeleton for more tests in the future. To run the tests, `pip install pytest` and then run `pytest` in the repo dir.

**Comment by sap218** [comment]
@majensen Thank you so much! I just merged the PR - I appreciate the help! :-) 

@balhoff Thank you!

@wdduncan We now have automated testing :-)

**Comment by majensen** [comment]
@wdduncan there are two boxes unchecked on your review . I think we've taken care of "Automated Tests". The last one is "State of the Field". If you're happy with where we are on these, can you confirm by checking those off. Then I can push our recommendation to the editors-in-chief.

**Comment by wdduncan** [comment]
@majensen they're checked now :)

**Comment by majensen** [comment]
@sap218 we're on the final approach. Can you create a tag for your master branch, and then use that to create a public archive for your software? You can use [Zenodo](https://zenodo.org/) -- there's a quick tutorial at https://github.com/openjournals/joss-reviews/issues/1839#issuecomment-560048729
That process will give you a DOI (like "10.5281/zenodo.3558815" ) - please report that back here. Thanks!

**Comment by sap218** [comment]
@majensen Done! 

10.5281/zenodo.3922261
[![DOI](https://zenodo.org/badge/227571502.svg)](https://zenodo.org/badge/latestdoi/227571502)

**Comment by majensen** [comment]
Excellent @sap218. Now, JOSS prefers that the name of the archive be the same as that of the paper, and that the author lists match. So.. would you be able to update the name and remove that parasitic author from the list on the archive?

**Comment by sap218** [comment]
@majensen I believe i have fixed it!

**Comment by majensen** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02168/joss.02168/10.21105.joss.02168.pdf)

**Comment by majensen** [comment]
@whedon set 10.5281/zenodo.3922261 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.3922261" target="_blank">10.5281/zenodo.3922261</a> is the archive.

**Comment by majensen** [comment]
@whedon set v1.0.0.0 as version

**Comment by whedon** [comment]
OK. v1.0.0.0 is the version.

**Comment by majensen** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.1093/database/bau033 is OK
- 10.3115/v1/p14-5010 is OK
- 10.1093/nar/gkr972 is OK
- 10.1016/j.ajhg.2008.09.017 is OK
- 10.1093/bib/bbv011 is OK

MISSING DOIs

- https://doi.org/10.25080/majora-92bf1922-00a may be missing for title: Data structures for statistical computing in python

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:wave: @openjournals/joss-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/1523

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/1523, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by majensen** [comment]
Hi @sap218 - can you check on the reference that whedon found missing in https://github.com/openjournals/joss-reviews/issues/2168#issuecomment-651877564 ?

**Comment by sap218** [comment]
@majensen just fixed! :-)

**Comment by danielskatz** [comment]
@majensen - please let me know when this is really ready to go

**Comment by majensen** [comment]
sure @danielskatz , although you might tell @whedon (or @arfon) not to summon the eic until it performs a dry run free of errors. Sorry to disturb.

**Comment by sap218** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.1093/database/bau033 is OK
- 10.3115/v1/p14-5010 is OK
- 10.1093/nar/gkr972 is OK
- 10.1016/j.ajhg.2008.09.017 is OK
- 10.1093/bib/bbv011 is OK
- 10.25080/majora-92bf1922-00a is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by majensen** [comment]
@openjournals/joss-eics really ready, thanks ever so much.

**Comment by danielskatz** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02168/joss.02168/10.21105.joss.02168.pdf)

**Comment by danielskatz** [comment]
👋 @sap218 - I have some comments/requests on the paper, in order from start to end:
- "gaining the key textual elements" - I don't know what this means; how does one gain textual elements?  Could "gaining" be "determining"?
- "having knowledge of the synonyms for clinical terminology may increase the size of the corpus available for analysis" - how is this the case?  How does having synonyms increase the size of the corpus?
- "e.g. social" <- "e.g., social"
- "their chosen ontology and" <- "their chosen ontology, and"
- "(Richardson, 2007)) and then" <- "(Richardson, 2007)), and then"
- "(e.g. lines from a corpus)" <- "(e.g., lines from a corpus)"
- "rank the important terms: using the" <- "rank the important terms by using the"
- "command arise will" <- "command, arise, will"
- "In Jabberwocky’s test repository (see Jabberwocky repo for further instructions), I show examples of each command separately. In the ‘process’ directory, I combine all three commands to demonstrate an example workflow. With 24 blog posts, the first use of catch returned 11 posts with the provided keywords. With bite I reviewed the CSV of ranked terms and curated new synonyms, simply by adding the corresponding class label from the ontology. I then used arise to add the identified synonyms into the ontology." <- "Jabberwocky’s test repository (see Jabberwocky repo for further instructions) shows examples of each command separately. The ‘process’ directory shows an example that combines all three commands to demonstrate an example workflow. With 24 blog posts, the first use of catch returned 11 posts with the provided keywords. The example uses bite to review the CSV of ranked terms and curated new synonyms, simply by adding the corresponding class label from the ontology. It then uses arise to add the identified synonyms into the ontology."
- For the reference Honnibal & Montani, "To appear" is not a journal. Please change this to include the journal name.
- For Manning et al., please protect cases with {}s in the bibtex so that the proceedings title appears in Title Case, rather than Sentence case. Also protect the {B} in the title so that it appears as Bloom, not bloom. Also Spacy should be spaCy, I think.
- In McKinney et al., do the same. Also protect the {P} in the title so that it appears as Python, not python. Also remove pdfs.semanticscholar.org
- In Pedregosa et al., fix Python in the title from python
- For Richardson, please add a URL or venue

When these are done, regenerate the pdf with ```@whedon generate pdf``` and let me know.

**Comment by sap218** [comment]
@wheden generate pdf

**Comment by sap218** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02168/joss.02168/10.21105.joss.02168.pdf)

**Comment by sap218** [comment]
@whedon check references

**Comment by sap218** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.1093/database/bau033 is OK
- 10.3115/v1/p14-5010 is OK
- 10.1093/nar/gkr972 is OK
- 10.1016/j.ajhg.2008.09.017 is OK
- 10.1093/bib/bbv011 is OK
- 10.25080/majora-92bf1922-00a is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by sap218** [comment]
@danielskatz I have updated the paper and addressed your comments

**Comment by danielskatz** [comment]
Thanks -  I think one was missed, however: the proceedings title for Manning et al. should be in Title Case, not Sentence case.

**Comment by sap218** [comment]
@danielskatz Ah yes sorry! Just fixed that now :-)

**Comment by danielskatz** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02168/joss.02168/10.21105.joss.02168.pdf)

**Comment by danielskatz** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.1093/database/bau033 is OK
- 10.3115/v1/p14-5010 is OK
- 10.1093/nar/gkr972 is OK
- 10.1016/j.ajhg.2008.09.017 is OK
- 10.1093/bib/bbv011 is OK
- 10.25080/majora-92bf1922-00a is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:wave: @openjournals/joss-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/1526

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/1526, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by danielskatz** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1278318856734466049) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/1527
1. Wait a couple of minutes to verify that the paper DOI resolves [https://doi.org/10.21105/joss.02168](https://doi.org/10.21105/joss.02168)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? Notify your editorial technical team...

**Comment by danielskatz** [comment]
Thanks to @wdduncan & @balhoff for reviewing!
And to @majensen for editing!

Congratulations to @sap218 (Samantha Pendleton) and co-author!!

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.02168/status.svg)](https://doi.org/10.21105/joss.02168)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.02168">
  <img src="https://joss.theoj.org/papers/10.21105/joss.02168/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.02168/status.svg
   :target: https://doi.org/10.21105/joss.02168
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.02168/status.svg)](https://doi.org/10.21105/joss.02168)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-joss

**Comment by majensen** [comment]
@balhoff and @wdduncan thanks very much from me personally - 
@sap218 : congratulations! Go get a kebab!

**Comment by wdduncan** [comment]
@majensen Thanks Mark!
Congrats @sap218 !!

**Comment by sap218** [comment]
@majensen @danielskatz @balhoff @wdduncan thank you all! I appreciate the help!
